### PR TITLE
feat: add installed plugin architecture

### PR DESCRIPTION
**Pull request by abhiksark**
## Summary

- add the Qt-free `labelimgplusplus.plugins` API-major-1 public contract and metadata-only PyPA entry-point discovery
- add transactional activation, namespaced JSON settings, generation-scoped background tasks, read-only document descriptors, structured diagnostics, and reverse teardown
- add host-owned plugin actions, retained dynamic shortcuts, restart-only enablement, safe mode, and the Tools → Plugins manager
- add a separately built fixture wheel, Python 3.8–3.13 focused CI, performance/teardown gates, PyInstaller boot coverage, and complete authoring/API/troubleshooting documentation

Refs #26.

#27 and #28 remain open and are blocked on this foundation being merged and documented.

## Test plan

- [x] `PYTHONPATH=. QT_QPA_PLATFORM=offscreen pytest -q` — 919 passed, 1 skipped
- [x] combined SAM + video matrix — 130 passed, 1 skipped
- [x] focused packaged plugin install/activate/execute/disable/uninstall lifecycle
- [x] Ruff on all changed Python inputs
- [x] compileall and Python 3.8 AST parsing
- [x] wheel/sdist content and dependency inspection
- [x] clean PyInstaller one-file artifact boot in plugin safe mode with no optional-feature hooks
- [x] five-run gates: no-plugin startup 10.80 ms p95; 100-entry discovery 0.19 ms p95; 100-row manager 2.70 ms p95; cancellation acknowledgement 1.10 ms p95; clean teardown

## Checklist

- [x] Tests added and passing
- [x] Documentation and HISTORY updated
- [x] Base dependencies and CLI contracts unchanged
- [x] Existing image/video formats, filenames, class ordering, synchronous APIs, and Qt ownership boundaries preserved
- [ ] Cross-version GitHub Actions pending
- [ ] Mypy not configured in this repository

## Summary by CodeRabbit

* **New Features**
  * Added support for installing and managing Python plugins through **Tools → Plugins…**.
  * Plugins can provide commands, shortcuts, settings, document access, background tasks, and diagnostics.
  * Added restart-required activation, safe mode, and an option to disable plugins for recovery.
  * Plugin settings are stored in JSON with improved retention and reset options.

* **Documentation**
  * Added plugin authoring guidance, API reference, architecture details, troubleshooting information, and release documentation.

* **Tests**
  * Expanded validation for plugin discovery, packaging, lifecycle behavior, performance, and recovery scenarios.

### CODE REVIEW

File: libs/core/plugin_manager.py [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+            )
+        if not metadata_value.display_name or not metadata_value.version:
+            raise _ValidationFailure(
+                "invalid_metadata", "Display name and version must be non-empty.")
+        if not isinstance(metadata_value.capabilities, tuple):
+            raise _ValidationFailure(
+                "invalid_metadata", "Capabilities must be a tuple.")
+        supported = {PluginCapability.COMMANDS}
+        unknown = [capability for capability in metadata_value.capabilities
+                   if capability not in supported]
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply libs/core/plugin_manager.py]: Fixed in the current five-commit head. Activation now validates API-major, metadata text fields, the capabilities tuple, and every PluginCapability value before compatibility checks;  covers string-enum and non-string metadata cases.

File: libs/core/plugin_registry.py [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+            self._manager._add_document_subscription(self._record, callback)
+            self._committed.append(callback)
+        self._staged = []
+
+    def close(self):
+        for callback in list(self._committed):
+            self._manager._remove_document_subscription(self._record, callback)
+        self._committed = []
+        self._staged = []
+        self._closed = True
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply libs/core/plugin_registry.py:209]: Fixed. The manager now retains one entry per subscription registration, so duplicate callbacks unsubscribe independently;  verifies two deliveries, then one, then none.

File: libs/widgets/pluginManagerDialog.py [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+            if diagnostic.details:
+                lines.append(diagnostic.details.rstrip())
+        self.diagnostics.setPlainText("\n\n".join(lines))
+        metadata = record.metadata
+        if metadata and metadata.homepage:
+            escaped = metadata.homepage.replace('"', "%22")
+            self.homepage.setText(
+                '<a href="%s">%s</a>' % (escaped, metadata.homepage))
+        else:
+            self.homepage.clear()
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply libs/widgets/pluginManagerDialog.py:369]: Fixed. Homepage links now require an  or  URL with a network location and HTML-escape both href and visible text. The integration test rejects  and verifies quote/tag escaping.

File: libs/widgets/pluginManagerDialog.py [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+        self._assert_gui_thread()
+        if self._enabled is None:
+            return
+        for command_id, action in tuple(self._actions.items()):
+            action.setEnabled(bool(self._enabled(command_id)))
+
+    def forget(self, plugin_id):
+        self._shortcut_config.forget_plugin(plugin_id)
+        self._settings[SETTING_SHORTCUTS] = self._shortcut_config.to_dict()
+        self._settings.save()
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply libs/widgets/pluginManagerDialog.py]: Fixed. Forgetting now participates in one settings save, returns/propagates failure, restores in-memory settings and shortcut state when persistence fails, and the dialog displays a warning. The failure path retains the unavailable record, configuration, and shortcut in its integration test.

File: libs/core/plugin_registry.py [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+            generation=generation,
+        )
+        bridge = _TaskBridge(self, handle, generation)
+        bridge.configure(on_result, on_error, on_progress)
+        self._bridges[handle] = bridge
+        handle.result.connect(bridge.result, Qt.QueuedConnection)
+        handle.error.connect(bridge.error, Qt.QueuedConnection)
+        handle.progress.connect(bridge.progress, Qt.QueuedConnection)
+        handle.finished.connect(bridge.finished, Qt.QueuedConnection)
+        return handle
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply libs/core/plugin_registry.py:357]: Fixed. Plugin work is held behind a short readiness event until all result/error/progress/finished connections are installed.  covers an immediate worker progress report.

File: libs/core/shortcut_config.py [hunk 6]
```diff
@@ -68,30 +69,77 @@ def get(self, action_name):
...
+
+    def forget_plugin(self, plugin_id):
+        """Remove defaults and retained overrides for one forgotten plugin."""
+        prefix = 'plugin.%s.' % plugin_id
+        for name in tuple(self._dynamic_defaults):
+            if name.startswith(prefix):
+                self._dynamic_defaults.pop(name, None)
+        for name in tuple(self._shortcuts):
+            if name.startswith(prefix):
+                self._shortcuts.pop(name, None)
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply libs/core/shortcut_config.py]: Fixed without ambiguous prefix parsing.  tracks and persists exact command-to-plugin ownership, and forgetting  preserves ; the dot-prefix regression is covered in both shortcut and host integration tests.

File: libs/core/plugin_registry.py [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+            self._service._invoke_callback(
+                "progress", self._progress_callback, value)
+
+    @pyqtSlot()
+    def finished(self):
+        if self._service is None:
+            return
+        self._service._finished(self._handle)
+        self.detach()
+        self.deleteLater()
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply libs/core/plugin_registry.py:279]: Fixed. TaskBridge delivery now rejects detached service or handle state. The shutdown regression queues progress before detach, captures sys.excepthook, processes queued Qt events, and asserts no unexpected exception.

File: libs/core/shortcut_config.py [hunk 8]
```diff
@@ -68,30 +69,77 @@ def get(self, action_name):
...
+        self._shortcuts.update(self._dynamic_defaults)
 
     def get_all(self):
         """Returns a copy of all current shortcuts."""
-        return dict(self._shortcuts)
+        visible = dict(DEFAULT_SHORTCUTS)
+        for name in self._dynamic_defaults:
+            visible[name] = self._shortcuts.get(
+                name, self._dynamic_defaults[name])
+        return visible
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply libs/core/shortcut_config.py:144]: Fixed. get_all now overlays stored built-in values and only exposes registered dynamic commands. The regression verifies the customized binding conflicts and the replaced default no longer does.
